doc/notes/plutus-tx-plugin/NO_SIMPLE_PASS.md

```diff
@@ -0,0 +1,135 @@
+# Investigation: Disabling mkSimplPass in PlutusTx Plugin
+
+## Context
+
+This investigation was conducted to understand whether disabling the `mkSimplPass` in the PlutusTx plugin would improve error messages for stage violations, specifically the confusing "Cannot construct a value of type: BuiltinUnit" error reported in [plutus-private#1626](https://github.com/IntersectMBO/plutus-private/issues/1626).
+
+## Background
+
+The `mkSimplPass` was originally added to work around [GHC #16615](https://gitlab.haskell.org/ghc/ghc/-/issues/16615), where local bindings lack unfoldings when Core plugins run early in the compilation pipeline. The simplifier pass with `sm_pre_inline = True` generates these unfoldings by inlining bindings that occur exactly once unconditionally.
+
+### Why mkSimplPass Was Needed
+
+From Note [GHC.sm_pre_inline] in the code:
+- The plugin requires certain functions to be fully applied
+- Example: it can handle `noinline @(String -> BuiltinString) stringToBuiltinString "a"` but not `let f = noinline @(String -> BuiltinString) stringToBuiltinString in f "a"`
+- Pre-inlining solves this by inlining single-use bindings before the plugin runs
+
+## Investigation Method
+
+1. Disabled `mkSimplPass` in `PlutusTx.Plugin.install` 
+2. Attempted to build the plutus-tx-plugin test suite
+3. Categorized and analyzed all compilation failures
+4. Assessed whether failures could be fixed without the simplifier pass
+
+## Key Findings
+
+### ✅ Primary Goal Achieved: Better Error Messages
+
+**Before (with mkSimplPass):**
+```
+Error: Unsupported feature: Cannot construct a value of type: BuiltinUnit
+Context: Compiling expr: BuiltinUnit
+Context: Compiling expr: BuiltinUnit ipv
+```
+
+**After (without mkSimplPass):**
+```
+Error: Reference to a name which is not a local, a builtin, or an external INLINABLE function: Variable validator1
+No unfolding
+Context: Compiling expr: validator1
+Context: Compiling expr at: test/BuiltinUnit/Spec.hs:74:11-36
+```
+
+The new error message clearly indicates the stage violation problem and points to the exact location, making it much easier for users to understand and fix.
+
+### Failure Categories
+
+Without `mkSimplPass`, we observed three main categories of failures:
+
+#### 1. Stage Violations (✅ Fixed)
+- **BuiltinUnit test cases**: Now produce clear "No unfolding" errors instead of confusing builtin constructor errors
+- **This is the desired behavior** - code that violates PlutusTx compilation rules should fail with meaningful messages
+
+#### 2. Missing Unfoldings for INLINEABLE Functions (🔧 Potentially Fixable)
+```
+Error: Reference to a name which is not a local, a builtin, or an external INLINABLE function: Variable Plugin.Coverage.Spec.fun
+OtherCon []
+```
+- **Root cause**: GHC #16615 - local bindings lack unfoldings without simplifier
+- **Pattern**: Functions marked `{-# INLINEABLE #-}` but defined after their use site
```

Comment:
It's probably also the case for function marked inlineable but defined on the other module.